source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,617 @@
+    if (TailBegin != AllocInfo->UserEnd) {
+        auto Value = AllocInfo->UserEnd -
+                     RoundDownTo(AllocInfo->UserEnd, ASAN_SHADOW_GRANULARITY);
+        UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->UserEnd,
+                                    1, Value, LastEvent, &LastEvent));
+    }
+
+    int ShadowByte;
+    switch (AllocInfo->Type) {
+    case USMMemoryType::HOST:
+        ShadowByte = kUsmHostRedzoneMagic;
+        break;
+    case USMMemoryType::DEVICE:
+        ShadowByte = kUsmDeviceRedzoneMagic;
+        break;
+    case USMMemoryType::SHARE:
+        ShadowByte = kUsmSharedRedzoneMagic;
+        break;
+    case USMMemoryType::MEM_BUFFER:
+        ShadowByte = kMemBufferRedzoneMagic;
+        break;
+    default:
+        ShadowByte = 0xff;
+        assert(false && "Unknow AllocInfo Type");
+    }
+
+    // Left red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
+                                AllocInfo->UserBegin - AllocInfo->AllocBegin,
+                                ShadowByte, LastEvent, &LastEvent));
+
+    // Right red zone
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, TailBegin,
+                                TailEnd - TailBegin, ShadowByte, LastEvent,
+                                &LastEvent));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::updateShadowMemory(ur_queue_handle_t Queue) {
+    auto Context = getContext(Queue);
+    auto Device = getDevice(Queue);
+    assert(Device != nullptr);
+
+    auto ContextInfo = getContextInfo(Context);
+
+    auto HostInfo = ContextInfo->getDeviceInfo(nullptr);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::shared_lock<ur_shared_mutex> HostGuard(HostInfo->Mutex,
+                                                std::defer_lock);
+    std::unique_lock<ur_shared_mutex> DeviceGuard(DeviceInfo->Mutex,
+                                                  std::defer_lock);
+    std::scoped_lock<std::shared_lock<ur_shared_mutex>,
+                     std::unique_lock<ur_shared_mutex>, ur_mutex>
+        Guard(HostGuard, DeviceGuard, QueueInfo->Mutex);
+
+    ur_event_handle_t LastEvent = QueueInfo->LastEvent;
+
+    // FIXME: Always update host USM, but it'd be better to update host USM
+    // selectively, or each devices once
+    for (auto &AllocInfo : HostInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+
+    for (auto &AllocInfo : DeviceInfo->AllocInfos) {
+        UR_CALL(enqueueAllocInfo(Context, Device, Queue, AllocInfo, LastEvent));
+    }
+    DeviceInfo->AllocInfos.clear();
+
+    QueueInfo->LastEvent = LastEvent;
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::addContext(ur_context_handle_t Context) {
+    auto ContextInfo = std::make_shared<ur_sanitizer_layer::ContextInfo>();
+
+    // Host Device
```

Comment:
Is this comment accurate? SYCL doesn't have a host device.
I suppose you meant just Host?
However, the code below `DeviceInfo->Type = DeviceType::CPU` is showing a CPU device is initialized.

If the intent is to initialize information for the host, do we currently have logic to handle the coordination between host and a device? I suggest to to remove below code since it is for the future. When the coordination is in place, we can add below code back.

Reply:
> I suppose you meant just Host?

Yes. I don't think host device is a wrong term. Host code runs on the host device.

> If the intent is to initialize information for the host, do we currently have logic to handle the coordination between host and a device?

Yes, we have. I have saved all host USM allocation information in the host device.

Reply:
SYCL doesn't have host device. What you meant is just "host". Please don't use "host device".

> Yes, we have. I have saved all host USM allocation information in the host device.

Could you please point me to the code where you saved the info?

Reply:
https://github.com/oneapi-src/unified-runtime/blob/79998b965232ff40e3a0b3829dd57f5466eb23b3/source/loader/layers/sanitizer/asan_interceptor.cpp#L91-L92

Reply:
according to offline sync, @AllanZyne would you remove this part of code?

Reply:
Done